# Capstone 2 – Project Proposal – Somatic Mutation

# 1 Objective: 

The goal is to segment different types of Somatic Germline Mutations in human genes associated with inherited and acquired diseases. It will be a one-stop-shop comprehensive collection of mutation data(Segments) for easy discovery in the era of personalized medicine. As part of this project I would like to find: 

a.	Somatic Germline segmentation

b.	Acquired Disease with maximum number of Somatic Mutation 

c.	Inherited Disease with maximum number of Somatic Mutation

### Outcome: Comprehensive Somatic Mutation Database an invaluable resource for all scientists. 

# 2 Client

#### Will be provided by Sona
Sample: The client for this project is Georgetown University (www.gwu.edu ) and the Bioinformatics Department. The purpose is to find an ML model which can be used to correctly cluster somatic mutation.

# 3 Data source and Credits


We will provided by Sona.

# 4. Solution Approach

The solution plans to use PCA and NMF ML techniques to help with dimension reduction and segmentation of Somatic Mutation. It is specifically designed to address large dataset with biodiversity and quality issues like redundancy, missing, wrong label etc. The solution is sub-divided into three phases as listed below.





### a)Data Assembly - Phase I: 
This phase of the project is designed to gather and do basic cleanup like join, merge, add or update attributes.

### b)Explore and Preprocessing – Phase II: 

This phase of the project is designed to validate and explore the dataset for all the problems listed in the “Problem” section of this proposal. 
### c)Modelling and Evaluation Phase III: 

In this phase of the project will focus on exploring various machine learning algorithms and finding the right hyperparameters to find the best ML model to cluster the Somatic Mutation.  

## 4.1 Data Assembly - Phase I:

OMIM Data identifier is O

UNIPROT data identifier is U

COSMIC data identifier is C
#import warnings
#warnings.simplefilter("ignore", DeprecationWarning)

In [1]:
import pandas as pd
from google.cloud import storage
client = storage.Client()
bucket=client.get_bucket('somatic_germline_mutations')
blob = storage.Blob('mim2gene.txt',bucket)
with open('mim2gene.txt', 'wb') as file_obj:
    blob.download_to_file(file_obj)
df_O=pd.read_csv('mim2gene.txt',sep='\t',skiprows=5, header=None, \
                names=['MIMNumber_O','MIMEntryType_O','EntrezGeneID_NCBI_O','GeneName','EnsemblGeneID_O'])
df_O=df_O[['GeneName','MIMNumber_O','MIMEntryType_O','EntrezGeneID_NCBI_O','EnsemblGeneID_O']] # reordering the columns
df_O.shape

(25471, 5)

### Number of entires in above OMIM list: 25,471 with 5 columns.

We are focusing on those entries which are associted with disease.

In [2]:
df_O=df_O[df_O.GeneName.notnull()] #Removing entries from omim file without a gene name.
df_O.shape

(16021, 5)

### Number of entires in above OMIM list after removing entries without gene name: 16,021 with 5 columns.

## Testing for Gene involved in multiple disorder

In [3]:
df_O[df_O.duplicated(subset='GeneName',keep=False)]

GeneName  MIMNumber_O MIMEntryType_O  EntrezGeneID_NCBI_O  \
2092       IGH       146910           gene               3492.0   
2107       IGH       147010           gene               3492.0   
2114       IGH       147070           gene               3492.0   
6280      ASMT       300015           gene                438.0   
6297      ATRX       300032           gene                546.0   
6416   SLC25A6       300151           gene                293.0   
6427     ASMTL       300162           gene               8623.0   
6554    XAGE1E       300289           gene             653067.0   
6622     CRLF2       300357           gene              64109.0   
6769      ATRX       300504      phenotype                546.0   
6935    SPANXD       300670           gene              64648.0   
6936    SPANXD       300671           gene              64648.0   
7007    XAGE1B       300742           gene             653220.0   
7008    XAGE1B       300743           gene             653220.0   
7009    XAGE1E       300744           gene             653067.0   
7010    XAGE1E       300745           gene             653067.0   
7397    CSF2RA       306250           gene               1438.0   
7443     IL3RA       308385           gene               3563.0   
7571   AKAP17A       312095           gene               8227.0   
7602      SHOX       312865           gene               6473.0   
7619      CD99       313470           gene               4267.0   
7656       XGR       314705      phenotype               7501.0   
7681     ASMTL       400011           gene               8623.0   
7690      SHOX       400020           gene               6473.0   
7693     CRLF2       400023           gene              64109.0   
7717      ASMT       402500           gene                438.0   
7718   SLC25A6       403000           gene                293.0   
7723    CSF2RA       425000           gene               1438.0   
7726     IL3RA       430000           gene               3563.0   
7727      CD99       450000           gene               4267.0   
7728   AKAP17A       465000           gene               8227.0   
7735       XGR       489500           gene               7501.0   
18223   NBPF15       610414           gene             284565.0   
18439   PTPN20       610630           gene              26095.0   
18440   PTPN20       610631           gene              26095.0   
21814   NBPF15       614005           gene             284565.0   

       EnsemblGeneID_O  
2092               NaN  
2107               NaN  
2114               NaN  
6280   ENSG00000196433  
6297   ENSG00000085224  
6416   ENSG00000169100  
6427   ENSG00000169093  
6554   ENSG00000204382  
6622   ENSG00000205755  
6769   ENSG00000085224  
6935   ENSG00000196406  
6936   ENSG00000196406  
7007   ENSG00000204379  
7008   ENSG00000204379  
7009   ENSG00000204382  
7010   ENSG00000204382  
7397   ENSG00000198223  
7443   ENSG00000185291  
7571   ENSG00000197976  
7602   ENSG00000185960  
7619   ENSG00000002586  
7656               NaN  
7681   ENSG00000169093  
7690   ENSG00000185960  
7693   ENSG00000205755  
7717   ENSG00000196433  
7718   ENSG00000169100  
7723   ENSG00000198223  
7726   ENSG00000185291  
7727   ENSG00000002586  
7728   ENSG00000197976  
7735               NaN  
18223  ENSG00000266338  
18439  ENSG00000204179  
18440  ENSG00000204179  
21814  ENSG00000266338

### Import Uniprot Data.

In [4]:
blob = storage.Blob('uniprot-organismHomosapiens9606.tab',bucket)
with open('uniprot-organismHomosapiens9606.tab', 'wb') as file_obj:
    blob.download_to_file(file_obj)

df_U=pd.read_csv('uniprot-organismHomosapiens9606.tab',sep='\t', header=0, \
               names=['Entry_U','ProteinName_U','GeneName','Organism_U','Entryname_U','EnsemblGeneID_U'])

df_U=df_U[['GeneName','Entry_U','ProteinName_U','Organism_U','Entryname_U','EnsemblGeneID_U']]
df_U.shape

(161521, 6)

In [5]:
df_U.head() 
#we can see that GeneName is comprising of more than one names. We are only considering the 1st Name as the GeneName

GeneName Entry_U  \
0              TP53 P53  P04637   
1            APP A4 AD1  P05067   
2                 SCN5A  Q14524   
3              FBN1 FBN  P35555   
4  EGFR ERBB ERBB1 HER1  P00533   

                                       ProteinName_U            Organism_U  \
0  Cellular tumor antigen p53 (Antigen NY-CO-13) ...  Homo sapiens (Human)   
1  Amyloid-beta A4 protein (ABPP) (APPI) (APP) (A...  Homo sapiens (Human)   
2  Sodium channel protein type 5 subunit alpha (H...  Homo sapiens (Human)   
3               Fibrillin-1 [Cleaved into: Asprosin]  Homo sapiens (Human)   
4  Epidermal growth factor receptor (EC 2.7.10.1)...  Homo sapiens (Human)   

   Entryname_U                                    EnsemblGeneID_U  
0    P53_HUMAN  ENST00000269305 [P04637-1];ENST00000420246 [P0...  
1     A4_HUMAN  ENST00000346798 [P05067-1];ENST00000348990 [P0...  
2  SCN5A_HUMAN  ENST00000333535 [Q14524-1];ENST00000423572 [Q1...  
3   FBN1_HUMAN                                   ENST00000316623;  
4   EGFR_HUMAN  ENST00000275493 [P00533-1];ENST00000342916 [P0...

### Number of entires in above Uniprot list: 161,521 with 6 columns.
We only started with 6 columns so that the uniprot list is managable. 

In [6]:
df_U=df_U[df_U.GeneName.notnull()] #Removing entries from Uniprot file without a gene name.
df_U['GeneName']=df_U['GeneName'].apply(lambda x:x.split(' ')[0])
df_U.shape

(139590, 6)

### Number of entires in above Uniprot list after removing entries without gene name: 139,590 with 6 columns.

In [7]:
df_merged=pd.merge(df_U,df_O,how='left', on='GeneName')
df_merged.shape

(139654, 10)

In [8]:
df_merged.head()

GeneName Entry_U                                      ProteinName_U  \
0     TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
1      APP  P05067  Amyloid-beta A4 protein (ABPP) (APPI) (APP) (A...   
2    SCN5A  Q14524  Sodium channel protein type 5 subunit alpha (H...   
3     FBN1  P35555               Fibrillin-1 [Cleaved into: Asprosin]   
4     EGFR  P00533  Epidermal growth factor receptor (EC 2.7.10.1)...   

             Organism_U  Entryname_U  \
0  Homo sapiens (Human)    P53_HUMAN   
1  Homo sapiens (Human)     A4_HUMAN   
2  Homo sapiens (Human)  SCN5A_HUMAN   
3  Homo sapiens (Human)   FBN1_HUMAN   
4  Homo sapiens (Human)   EGFR_HUMAN   

                                     EnsemblGeneID_U  MIMNumber_O  \
0  ENST00000269305 [P04637-1];ENST00000420246 [P0...     191170.0   
1  ENST00000346798 [P05067-1];ENST00000348990 [P0...     104760.0   
2  ENST00000333535 [Q14524-1];ENST00000423572 [Q1...     600163.0   
3                                   ENST00000316623;     134797.0   
4  ENST00000275493 [P00533-1];ENST00000342916 [P0...     131550.0   

  MIMEntryType_O  EntrezGeneID_NCBI_O  EnsemblGeneID_O  
0           gene               7157.0  ENSG00000141510  
1           gene                351.0  ENSG00000142192  
2           gene               6331.0  ENSG00000183873  
3           gene               2200.0  ENSG00000166147  
4           gene               1956.0  ENSG00000146648

### Number of entires after merging OMIM and Uniprot list : 139,654 with 10 columns.

We are now focusing on Genes having known MIM IDs(Genes known to have involment in disease). We have filtered out the entires without any MIM ID.

In [9]:
df_merged=df_merged.dropna(subset=['MIMNumber_O'])
df_merged.shape
#df_merged[df_merged.MIMNumber_O.notnull()]
#df_O=df_O.dropna(subset=['GeneName'])
#df_merged[df_merged['GeneName'] =='IGH']

(106097, 10)

# Number of entires in the merged list after removing entries without MIM ids: 106,097 with 10 columns.

In [10]:
df_merged[df_merged['GeneName'] =='TP53']#.to_csv('xx.csv') # checking one sample gene

GeneName     Entry_U  \
0          TP53      P04637   
21697      TP53  A0A218MJD5   
23178      TP53      K7PPA8   
23835      TP53  A0A1B1PFD4   
24022      TP53      A2I9Y7   
24079      TP53      S5LJ61   
24225      TP53      A4GW67   
29721      TP53      Q1HGV3   
29855      TP53      L0EQE2   
30088      TP53      L0ES54   
30115      TP53      L0EPV7   
30471      TP53      A2I9Z0   
30800      TP53      A2I9Z2   
31132      TP53  A0A218MJF3   
32049      TP53      L0EQ92   
32111      TP53      S5LQX7   
32481      TP53      L0EQS1   
33671      TP53      L0ESB7   
34087      TP53      S5M706   
34113      TP53      Q1MSW9   
34413      TP53      L0EQP9   
35965      TP53  A0A1B1PFC7   
36436      TP53      A4GWB5   
36532      TP53      L0EQU5   
36971      TP53      Q1MSW8   
37116      TP53      L0ENG9   
38928      TP53      L0EQ05   
39043      TP53      L0EPY0   
39236      TP53      U6BHT9   
39329      TP53      Q1MSX0   
...         ...         ...   
102707     TP53  A0A087WXZ1   
102708     TP53  A0A087WZU8   
103047     TP53      B5AKF5   
103142     TP53      A4GW75   
103247     TP53      D5KL87   
104907     TP53      L0EQ03   
105139     TP53      L0ES73   
106158     TP53      L0EQ52   
107912     TP53      B4XAK8   
109019     TP53      U6BHU6   
109462     TP53      L0EP82   
110221     TP53      A2I9Z1   
111319     TP53      D5KL89   
111515     TP53      Q19RW6   
113931     TP53      L0EQV3   
114937     TP53      L0ENF7   
116785     TP53      S5LME3   
116983     TP53      U6BGT3   
118568     TP53      E5RMA8   
120202     TP53      L0EQT1   
121526     TP53  A0A218MJD9   
121973     TP53      Q19RW5   
127231     TP53      L0ERI9   
129344     TP53      E9PCY9   
131771     TP53      S5M207   
135224     TP53      B1PZ15   
135485     TP53      E9PFT5   
136754     TP53      L0EQX3   
136801     TP53      Q0ZAK0   
137180     TP53      L0EQJ1   

                                            ProteinName_U  \
0       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
21697               Cellular tumor antigen p53 (Fragment)   
23178                          Cellular tumor antigen p53   
23835                                      p53 (Fragment)   
24022                  Tumor protein p53 (p53) (Fragment)   
24079               Cellular tumor antigen p53 (Fragment)   
24225               Cellular tumor antigen p53 (Fragment)   
29721              Mutant p53 tumor suppressor (Fragment)   
29855                     Tumor suppressor p53 (Fragment)   
30088                     Tumor suppressor p53 (Fragment)   
30115               Cellular tumor antigen p53 (Fragment)   
30471                  Tumor protein p53 (p53) (Fragment)   
30800                  Tumor protein p53 (p53) (Fragment)   
31132               Cellular tumor antigen p53 (Fragment)   
32049               Cellular tumor antigen p53 (Fragment)   
32111               Cellular tumor antigen p53 (Fragment)   
32481               Cellular tumor antigen p53 (Fragment)   
33671                     Tumor suppressor p53 (Fragment)   
34087               Cellular tumor antigen p53 (Fragment)   
34113               Cellular tumor antigen p53 (Fragment)   
34413               Cellular tumor antigen p53 (Fragment)   
35965                                      p53 (Fragment)   
36436               Cellular tumor antigen p53 (Fragment)   
36532               Cellular tumor antigen p53 (Fragment)   
36971               Cellular tumor antigen p53 (Fragment)   
37116               Cellular tumor antigen p53 (Fragment)   
38928               Cellular tumor antigen p53 (Fragment)   
39043               Cellular tumor antigen p53 (Fragment)   
39236                                     TP53 (Fragment)   
39329               Cellular tumor antigen p53 (Fragment)   
...                                                   ...   
102707                         Cellular tumor antigen p53   
102708                         Cellular tumor antigen p53   
103047    

### Assumption and Consideration:
Please note that Uniprot is considered as the master list and merge with the OMIM list.

We can see that the numner of entries increased from the original Uniprot. Possible reasons(need investigation):

1) We may have GeneName in OMIM not present Uniprot list

2) There can be duplicate entries

In [ ]:
blob = storage.Blob('CosmicCompleteTargetedScreensMutantExport.tsv',bucket)
with open('CosmicCompleteTargetedScreensMutantExport.tsv', 'wb') as file_obj:
    blob.download_to_file(file_obj)

## Note:
Import mutation dataset from COSMIC database. It provides a tab separated table of the complete curated COSMIC dataset (targeted screens) from the current release. It includes all coding point mutations, and the negative data set. The name of the file is 'CosmicCompleteTargetedScreensMutantExport.tsv.gz'

We will only consider those entires which have Mutation IDs.

In [11]:
import pandas as pd
chunksize = 10 ** 6
cosmic_C=pd.DataFrame()
colnames=["GeneName","AccessionNumber","GeneCDSlength","HGNCid","SampleName","SampleId", \
          "IdTumour", "PrimarySite","SiteSubtype1","SiteSubtype2","SiteSubtype3", \
          "PrimaryHistology","HistologySubtype1", "HistologySubtype2","HistologySubtype3", \
          "GenomeWideScreen","MutationId","MutationCDS","MutationAA", "MutationDescription", \
          "MutationZygosity","LOH","GRCh","MutationGenomePosition","MutationStrand", 
          "SNP","ResistanceMutation","FATHMMPrediction","FATHMMScore","MutationSomaticStatus", \
          "Pubmed_PMID", "IdStudy","SampleSource","TumourOrigin","Age"]

for chunk in pd.read_csv('CosmicCompleteTargetedScreensMutantExport.tsv',sep='\t',header=0, \
                         names=colnames,low_memory=False, \
                 dtype={"GeneName":object,"AccessionNumber":object,"GeneCDSlength":object, \
                        "HGNCid":object,"SampleName":object,"SampleId":object,"IdTumour":object, \
                        "PrimarySite":object,"SiteSubtype1":object,"SiteSubtype2":object, \
                        "SiteSubtype3":object,"PrimaryHistology":object,"HistologySubtype1":object, \
                        "HistologySubtype2":object,"HistologySubtype3":object,"GenomeWideScreen":object, \
                        "MutationId":object,"MutationCDS":object,"MutationAA":object, \
                        "MutationDescription":object, "MutationZygosity":object,"LOH":object,"GRCh":object, \
                        "MutationGenomePosition":object, "MutationStrand":object,"SNP":object, \
                        "ResistanceMutation":object,"FATHMMPrediction":object, 
                        "FATHMMScore":object,"MutationSomaticStatus":object,"Pubmed_PMID":object, \
                        "IdStudy":object,"SampleSource":object,"TumourOrigin":object,"Age":object}, \
                 chunksize=chunksize
                ):
    cosmic_C=chunk[['GeneName','AccessionNumber','MutationId','MutationDescription', \
                  'MutationGenomePosition']].loc[chunk.MutationId.notnull()]
                  #selecting only with mutation id from Cosmic file

In [12]:
cosmic_C.shape

(13155, 5)

In [13]:
cosmic_C.head()

GeneName  AccessionNumber   MutationId      MutationDescription  \
6000185     EGFR  ENST00000275493    COSM12979  Substitution - Missense   
6000186     EGFR  ENST00000275493    COSM13243      Deletion - In frame   
6000187   PIK3CA      NM_006218.1      COSM775  Substitution - Missense   
6000188    FGFR3  ENST00000440486    COSM35896  Substitution - Missense   
6000189     IDH2  ENST00000330062  COSM1685352                  Unknown   

        MutationGenomePosition  
6000185                    NaN  
6000186                    NaN  
6000187  3:179234297-179234297  
6000188                    NaN  
6000189                    NaN

## Number of entires in the COSMIC list after removing entires without mutation IDs: 13,155 with 5 columns.

In [14]:
df_merged=pd.merge(df_merged,cosmic_C,how='left', on='GeneName')
df_merged.shape

(503106, 14)

### Number of entires in the merged list after including COSMIC dataset: 503,106 with 14 columns.

In [15]:
df_merged[df_merged['GeneName'] =='TP53'] # checking one sample gene. We can see lot of duplicate entries in the list.

GeneName Entry_U                                      ProteinName_U  \
0          TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
1          TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
2          TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
3          TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
4          TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
5          TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
6          TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
7          TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
8          TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
9          TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
10         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
11         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
12         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
13         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
14         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
15         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
16         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
17         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
18         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
19         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
20         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
21         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
22         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
23         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
24         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
25         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
26         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
27         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
28         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
29         TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
...         ...     ...                                                ...   
495571     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495572     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495573     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495574     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495575     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495576     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495577     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495578     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495579     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495580     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495581     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495582     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495583     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495584     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495585     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495586     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495587     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495588     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495589     TP53  L0EQJ1              Cellular tumor antigen p53 (Fragment)   
495590     TP53  L0EQJ1      

In [16]:
df_merged.head()

GeneName Entry_U                                      ProteinName_U  \
0     TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
1     TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
2     TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
3     TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
4     TP53  P04637  Cellular tumor antigen p53 (Antigen NY-CO-13) ...   

             Organism_U Entryname_U  \
0  Homo sapiens (Human)   P53_HUMAN   
1  Homo sapiens (Human)   P53_HUMAN   
2  Homo sapiens (Human)   P53_HUMAN   
3  Homo sapiens (Human)   P53_HUMAN   
4  Homo sapiens (Human)   P53_HUMAN   

                                     EnsemblGeneID_U  MIMNumber_O  \
0  ENST00000269305 [P04637-1];ENST00000420246 [P0...     191170.0   
1  ENST00000269305 [P04637-1];ENST00000420246 [P0...     191170.0   
2  ENST00000269305 [P04637-1];ENST00000420246 [P0...     191170.0   
3  ENST00000269305 [P04637-1];ENST00000420246 [P0...     191170.0   
4  ENST00000269305 [P04637-1];ENST00000420246 [P0...     191170.0   

  MIMEntryType_O  EntrezGeneID_NCBI_O  EnsemblGeneID_O  AccessionNumber  \
0           gene               7157.0  ENSG00000141510  ENST00000269305   
1           gene               7157.0  ENSG00000141510  ENST00000269305   
2           gene               7157.0  ENSG00000141510  ENST00000269305   
3           gene               7157.0  ENSG00000141510  ENST00000269305   
4           gene               7157.0  ENSG00000141510  ENST00000269305   

  MutationId      MutationDescription MutationGenomePosition  
0  COSM43699  Substitution - Missense     17:7674194-7674194  
1  COSM44555  Substitution - Missense     17:7675176-7675176  
2  COSM11307  Substitution - Missense     17:7674888-7674888  
3  COSM43751                  Unknown     17:7674291-7674291  
4  COSM10991  Substitution - Missense     17:7675216-7675216

# SNP and Mutation from Uniprot

In [17]:
blob = storage.Blob('uniprotOrganismHomosapiens9606.tab',bucket)
with open('uniprotOrganismHomosapiens9606.tab', 'wb') as file_obj:
    blob.download_to_file(file_obj)
import pandas as pd
df_snp_u=pd.read_csv('uniprotOrganismHomosapiens9606.tab',sep='\t', header=0, \
               names=['Entry','GeneName','NaturalVariant','Mutagenesis'],
                    dtype={'Entry':object,'GeneName':object,'NaturalVariant':object,'Mutagenesis':object}
                    )   

In [18]:
df_snp_u.shape

(161521, 4)

In [19]:
df_snp_u=df_snp_u[df_snp_u.GeneName.notnull()] #Removing entries from Uniprot file without a gene name.
df_snp_u['GeneName']=df_snp_u['GeneName'].apply(lambda x:x.split(' ')[0])
df_snp_u.head()

Entry GeneName                                     NaturalVariant  \
0  P04637     TP53  VARIANT 5 5 Q -> H (in a sporadic cancer; soma...   
1  P05067      APP  VARIANT 501 501 E -> K (in dbSNP:rs45588932). ...   
2  Q14524    SCN5A  VARIANT 9 9 G -> V (in LQT3; dbSNP:rs199473043...   
3  P35555     FBN1  VARIANT 20 20 Y -> C (in MFS). {ECO:0000269|Pu...   
4  P00533     EGFR  VARIANT 30 297 Missing (variant EGFR vIII; fou...   

                                         Mutagenesis  
0  MUTAGEN 15 15 S->A: Loss of interaction with P...  
1  MUTAGEN 99 102 KRGR->NQGG: Reduced heparin-bin...  
2  MUTAGEN 1476 1476 Q->K: Induces accelerated re...  
3  MUTAGEN 1542 1542 G->D: Loss of integrin-media...  
4  MUTAGEN 275 275 Y->A: Strongly reduced autopho...

In [20]:
#Removing entries from omim file without a gene name.
df_snp_u=df_snp_u[(df_snp_u.NaturalVariant.notnull()) | (df_snp_u.Mutagenesis.notnull())] 
#df_snp_u.shape
df_snp_u.sample(17)

Entry GeneName                                     NaturalVariant  \
1293   Q9H0Y0    ATG10  VARIANT 62 62 S -> P (in dbSNP:rs3734114). /FT...   
3138   Q86Y33   CDC20B  VARIANT 8 8 T -> P (in dbSNP:rs173042). /FTId=...   
15591  P48061   CXCL12                                                NaN   
15318  Q8N2Y8    RUSC2  VARIANT 73 73 T -> A (in dbSNP:rs1535422). {EC...   
18539  Q13488   TCIRG1  VARIANT 56 56 R -> W (in dbSNP:rs36027301). /F...   
10638  A6NI61     MYMK  VARIANT 91 91 P -> T (in CFZS; unknown patholo...   
1870   Q8NEM8    AGBL3  VARIANT 45 45 F -> Y (in dbSNP:rs2348049). /FT...   
11394  Q9BZQ8  FAM129A  VARIANT 633 633 S -> L (in dbSNP:rs12750174). ...   
13093  A1L4L8  PLAC8L1  VARIANT 11 11 C -> S (in dbSNP:rs12187913). /F...   
15375  Q9HBV2   SPACA1  VARIANT 237 237 L -> S (in dbSNP:rs2276089). /...   
13027  O43272    PRODH  VARIANT 8 8 P -> L. /FTId=VAR_064883.; VARIANT...   
10265  Q9UBM8   MGAT4C  VARIANT 428 428 T -> S (in dbSNP:rs17855890). ...   
10205  Q9UGB7     MIOX                                                NaN   
3665   Q6PB30    CSAG1  VARIANT 28 28 Y -> F (in dbSNP:rs1894359). /FT...   
19170  Q9NQH7  XPNPEP3  VARIANT 450 450 I -> L (in dbSNP:rs17002243). ...   
11177  P19404   NDUFV2  VARIANT 29 29 V -> A (in dbSNP:rs906807). {ECO...   
14944  P23396     RPS3                                                NaN   

                                             Mutagenesis  
1293                                                 NaN  
3138                                                 NaN  
15591  MUTAGEN 22 23 Missing: Abolished CXCR4 activat...  
15318                                                NaN  
18539                                                NaN  
10638                                                NaN  
1870                                                 NaN  
11394                                                NaN  
13093                                                NaN  
15375                                                NaN  
13027                                                NaN  
10265                                                NaN  
10205  MUTAGEN 127 127 K->S: Strongly reduced enzyme ...  
3665                                                 NaN  
19170                                                NaN  
11177                                                NaN  
14944  MUTAGEN 6 6 S->A: No effect on phosphorylation...

In [21]:
df_snp_u[df_snp_u['GeneName'] =='TP53']

Entry GeneName                                     NaturalVariant  \
0  P04637     TP53  VARIANT 5 5 Q -> H (in a sporadic cancer; soma...   

                                         Mutagenesis  
0  MUTAGEN 15 15 S->A: Loss of interaction with P...

In [22]:
df_merged=pd.merge(df_merged,df_snp_u,how='left', on='GeneName')
df_merged.shape

(4464294, 17)

## Number of entires in the merged list after including SNP and Mutations from Uniprot dataset: 4,464,294 with 17 columns.¶

In [23]:
df_merged.columns=['GeneName', 'Entry_U', 'ProteinName_U', 'Organism_U', 'Entryname_U',
       'EnsemblGeneID_U', 'MIMNumber_O', 'MIMEntryType_O',
       'EntrezGeneID_NCBI_O', 'EnsemblGeneID_O', 'AccessionNumber_C',
       'MutationId_C', 'MutationDescription_C', 'MutationGenomePosition_C', 'UniprotID',
       'NaturalVariant_U', 'Mutagenesis_U']

In [24]:
df_merged=df_merged[['UniprotID','GeneName', 'Entry_U', 'ProteinName_U','EnsemblGeneID_U', \
                     'EnsemblGeneID_O','EntrezGeneID_NCBI_O','MIMNumber_O', 'MIMEntryType_O', \
                     'NaturalVariant_U','Mutagenesis_U','AccessionNumber_C','MutationId_C', \
                     'MutationDescription_C', 'MutationGenomePosition_C']]

In [25]:
df_merged[df_merged['GeneName'] =='TP53']

UniprotID GeneName Entry_U  \
0          P04637     TP53  P04637   
1          P04637     TP53  P04637   
2          P04637     TP53  P04637   
3          P04637     TP53  P04637   
4          P04637     TP53  P04637   
5          P04637     TP53  P04637   
6          P04637     TP53  P04637   
7          P04637     TP53  P04637   
8          P04637     TP53  P04637   
9          P04637     TP53  P04637   
10         P04637     TP53  P04637   
11         P04637     TP53  P04637   
12         P04637     TP53  P04637   
13         P04637     TP53  P04637   
14         P04637     TP53  P04637   
15         P04637     TP53  P04637   
16         P04637     TP53  P04637   
17         P04637     TP53  P04637   
18         P04637     TP53  P04637   
19         P04637     TP53  P04637   
20         P04637     TP53  P04637   
21         P04637     TP53  P04637   
22         P04637     TP53  P04637   
23         P04637     TP53  P04637   
24         P04637     TP53  P04637   
25         P04637     TP53  P04637   
26         P04637     TP53  P04637   
27         P04637     TP53  P04637   
28         P04637     TP53  P04637   
29         P04637     TP53  P04637   
...           ...      ...     ...   
4340023    P04637     TP53  L0EQJ1   
4340024    P04637     TP53  L0EQJ1   
4340025    P04637     TP53  L0EQJ1   
4340026    P04637     TP53  L0EQJ1   
4340027    P04637     TP53  L0EQJ1   
4340028    P04637     TP53  L0EQJ1   
4340029    P04637     TP53  L0EQJ1   
4340030    P04637     TP53  L0EQJ1   
4340031    P04637     TP53  L0EQJ1   
4340032    P04637     TP53  L0EQJ1   
4340033    P04637     TP53  L0EQJ1   
4340034    P04637     TP53  L0EQJ1   
4340035    P04637     TP53  L0EQJ1   
4340036    P04637     TP53  L0EQJ1   
4340037    P04637     TP53  L0EQJ1   
4340038    P04637     TP53  L0EQJ1   
4340039    P04637     TP53  L0EQJ1   
4340040    P04637     TP53  L0EQJ1   
4340041    P04637     TP53  L0EQJ1   
4340042    P04637     TP53  L0EQJ1   
4340043    P04637     TP53  L0EQJ1   
4340044    P04637     TP53  L0EQJ1   
4340045    P04637     TP53  L0EQJ1   
4340046    P04637     TP53  L0EQJ1   
4340047    P04637     TP53  L0EQJ1   
4340048    P04637     TP53  L0EQJ1   
4340049    P04637     TP53  L0EQJ1   
4340050    P04637     TP53  L0EQJ1   
4340051    P04637     TP53  L0EQJ1   
4340052    P04637     TP53  L0EQJ1   

                                             ProteinName_U  \
0        Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
1        Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
2        Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
3        Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
4        Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
5        Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
6        Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
7        Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
8        Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
9        Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
10       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
11       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
12       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
13       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
14       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
15       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
16       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
17       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
18       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
19       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
20       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
21       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
22       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
23       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
24       Cellular tumor antigen p53 (Antigen NY-CO-13) ...   
25       Cellular tumor antigen p53 (An

In [26]:
df_merged.columns

Index(['UniprotID', 'GeneName', 'Entry_U', 'ProteinName_U', 'EnsemblGeneID_U',
       'EnsemblGeneID_O', 'EntrezGeneID_NCBI_O', 'MIMNumber_O',
       'MIMEntryType_O', 'NaturalVariant_U', 'Mutagenesis_U',
       'AccessionNumber_C', 'MutationId_C', 'MutationDescription_C',
       'MutationGenomePosition_C'],
      dtype='object')

In [27]:
import datetime as dt
from datetime import datetime
from pytz import timezone

import uuid

tz = timezone('EST') # adding time zone info
datetime.now(tz) 
df_merged['Entrydate'] = dt.datetime.now()

df_merged.insert(0,'Id',uuid.uuid4()) 
df_merged.Id= df_merged.Id.apply(lambda x: uuid.uuid4()) # adding unique identifier

In [28]:
df_merged['EntrezGeneID_NCBI_O']=df_merged.EntrezGeneID_NCBI_O.apply(lambda x: str(x))
df_merged['MIMNumber_O']=df_merged.MIMNumber_O.apply(lambda x: str(x))

In [31]:
df_merged.head(200).to_csv('somaticFinal.csv')

# Inserting Somatic Data into Cassandra Database
The step is done for persistence and reliability among other benefits using a database cluster.

In [ ]:
import itertools
from multiprocessing import Pool
import sys
import time
from cassandra.cluster import Cluster
from cassandra.concurrent import execute_concurrent_with_args
from cassandra.query import tuple_factory
from cassandra.auth import PlainTextAuthProvider

In [ ]:
df_con=pd.read_csv('~/connection_point.csv',header=0) # this is done to make add basic level of security. 
#Please note that this file is not uploaded. It is only present in the Jupyter server. 

In [ ]:
def _insertData(params):
    cluster = Cluster(contact_points=[df_con.ip[0]], auth_provider = \
                      PlainTextAuthProvider(username=df_con.user[0], \
                                            password=df_con.token[0]))
    session = cluster.connect()
    session.set_keyspace('somatic')
    session.row_factory = tuple_factory
    prepared=session.prepare("INSERT INTO TABLE somatic.somaticMerged \
                             (id,UniprotID,GeneName,Entry_U,ProteinName_U, \
                             EnsemblGeneID_U,EnsemblGeneID_O,EntrezGeneID_NCBI_O, \
                             MIMNumber_O,MIMEntryType_O,NaturalVariant_U,Mutagenesis_U, \
                             AccessionNumber_C,MutationId_C,MutationDescription_C, \
                             MutationGenomePosition_C,entrydate) \
                             VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)")            
    
    #using datastax driver for multiprocessing 
    execute_concurrent_with_args(session, prepared, params, concurrency=50) 
    return None

def multiprocess(params):
    pool = Pool(processes=2)
    results = [pool.map(_insertData, (params[n:n+100],)) for n in range(0, len(params),100)]
    return results

if __name__ == "__main__":
    parameters=[]
    for index, row in enumerate(df_merged.values):        
        (a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q) = row
        row1=(a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q)
        parameters.append(row1)           
    a = multiprocess(parameters)